src/Cake.Core/IO/NuGet/NuGetPackageAssembliesLocator.cs

```diff
@@ -0,0 +1,96 @@
+    /// </summary>
+    public class NuGetPackageAssembliesLocator : INuGetPackageAssembliesLocator
+    {
+        private readonly INuGetAssemblyCompatibilityFilter _assemblyCompatibilityFilter;
+        private readonly ICakeEnvironment _environment;
+        private readonly IFileSystem _fileSystem;
+        private readonly ICakeLog _log;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NuGetPackageAssembliesLocator"/> class.
+        /// </summary>
+        /// <param name="fileSystem">The file system.</param>
+        /// <param name="log">The log.</param>
+        /// <param name="assemblyCompatibilityFilter">The assembly compatibility filter.</param>
+        /// <param name="environment">The environment.</param>
+        public NuGetPackageAssembliesLocator(IFileSystem fileSystem, ICakeLog log,
+            INuGetAssemblyCompatibilityFilter assemblyCompatibilityFilter, ICakeEnvironment environment)
+        {
+            _fileSystem = fileSystem;
+            _log = log;
+            _assemblyCompatibilityFilter = assemblyCompatibilityFilter;
+            _environment = environment;
+        }
+
+        /// <summary>
+        /// Finds assemblies (DLLs) included in a nuget package.
+        /// </summary>
+        /// <param name="packageDirectory">The addin directory.</param>
+        /// <returns>
+        /// the DLLs as IFile[].
+        /// </returns>
+        public IFile[] FindAssemblies(DirectoryPath packageDirectory)
+        {
+            if (packageDirectory == null)
+            {
+                throw new ArgumentNullException("packageDirectory");
+            }
+            if (packageDirectory.IsRelative)
+            {
+                throw new CakeException("Package directory (" + packageDirectory.FullPath + ") must be an absolute path.");
+            }
+
+            if (!_fileSystem.Exist(packageDirectory))
+            {
+                return new IFile[0];
+            }
+
+            var packageAssemblies = GetAllPackageAssemblies(packageDirectory);
+
+            if (!packageAssemblies.Any())
+            {
+                _log.Warning("Unable to locate any assemblies under {0}", packageDirectory.FullPath);
+            }
+
+            var compatibleAssemblies = FilterCompatibleAssemblies(packageAssemblies, packageDirectory);
+
+            return compatibleAssemblies.Select(ca => _fileSystem.GetFile(ca)).ToArray();
+        }
+
+        private FilePath[] GetAllPackageAssemblies(DirectoryPath packageDirectory)
+        {
+            return _fileSystem.GetDirectory(packageDirectory).GetFiles("*.dll", SearchScope.Recursive)
+                .Where(
+                    file =>
+                        !"Cake.Core.dll".Equals(file.Path.GetFilename().FullPath, StringComparison.OrdinalIgnoreCase))
+                .Select(a => a.Path)
+                .ToArray();
+        }
+
+        private FilePath[] FilterCompatibleAssemblies(
```

Comment:
Would comments be better, or just more descriptive/explicit method naming?
(I really hate to resort to comments when code can be used to do the job)